AppleSource/shell_cmds-198/chroot/chroot.8
.\" Copyright (c) 1988, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)chroot.8	8.1 (Berkeley) 6/9/93
.\"	$NetBSD: chroot.8,v 1.7 1998/10/06 03:47:50 mrg Exp $
.\"
.Dd October 6, 1998
.Dt CHROOT 8
.Os BSD 4.3
.Sh NAME
.Nm chroot
.Nd change root directory
.Sh SYNOPSIS
.Nm
.Op Fl u user
.Op Fl g group
.Op Fl G group,group,...
.Ar newroot
.Op Ar command
.Sh DESCRIPTION
The
.Nm
command changes its root directory to the supplied directory
.Ar newroot
and exec's
.Ar command ,
if supplied, or an interactive copy of your shell.
.Pp
If the
.Fl u ,
.Fl g
or
.Fl G
options are given, the user, group and group list of the process are
set to these values after the chroot has taken place.  See
.Xr setgid 2 ,
.Xr setgroups 2 ,
.Xr setuid 2 ,
.Xr getgrnam 3
and
.Xr getpwnam 3 .
.Pp
Note,
.Ar command
or the shell are run as your real-user-id.
.Sh ENVIRONMENT
The following environment variable is referenced by
.Nm :
.Bl -tag -width SHELL
.It Ev SHELL
If set,
the string specified by
.Ev SHELL
is interpreted as the name of
the shell to exec.
If the variable
.Ev SHELL
is not set,
.Pa /bin/sh
is used.
.El
.Sh SEE ALSO
.Xr chdir 2 ,
.Xr chroot 2 ,
.Xr environ 7
.Sh HISTORY
The
.Nm
utility first appeared in
.Bx 4.4 .
.Sh SECURITY CONSIDERATIONS
.Nm
should never be installed setuid root, as it would then be possible
to exploit the program to gain root privileges.


CHANGELOG.md
# Version 5.028.1
	- Fixed Cmd+v shortcut
	- HostKey fingerprint as base64 encoding
	- Smoothing the HUD on resize or SplitView
	- Empty default user configuration bugs.
	- Zoom shortcuts changed.
	- Solarized theme fixes.
	
# Version 4.024.2
	- AirTerminals! Put a terminal on your remote AirPlay screen :)
	- Blink shortcuts. Use your external keyboard to move, create, or remove terminals. Configure the trigger too! View settings > Keyboard > Shortcuts for more info.
	- Want to use SmartKeys when an external keyboard is connected? Now you can switch them on and off from configuration.
	- View geometry of the screen from the overlay with each resize.
	- Switch cursor blinking on/off from Appearance settings.

	- Fixed TouchID issues when returning to the app.
	- Fixed cursor blink sequence so that it always starts as ON.
	- Fixed Host Port not resetting properly.
	- Fixed overlay getting covered in landscape mode.

	- Smoother experience thanks to cleanups and improvements.

# Version 3.021.2
	- iCloud Hosts sync. Synchronize hosts between devices. If a Host already has been synced, it provides conflict resolution. No critical data like passwords is saved.
	- Auto Lock. If enabled, when you lock/unlock your device, Blink will also be locked. Passcode and TouchID will be required to unlock the app.
	- Added ARMv7 support. Support for 32 bit devices like iPad 2, 3, iPhone 5, etc... We will publish depending on how well it performs!
	- Added IPv6 support for hosts.
	- Share Public Encryption Keys. You can now share the public key from the Keys section to other apps, like Mail.

	- Updated Fira Code font to v1.204.
	- Improved error checking on Themes and Font uploads. Auto correct if the GH URL is not a raw one.

	- Fixed bug with password not getting saved on host creation.
	- Fixed crash when hitting arrows with landscape keyboard on Plus devices.
	- Rolled back LC_CTYPE enforcement on server.

# Version 2.109 / 3.0
	- NEW Autorepeat for normal keys on external keyboard. Vim users rejoice!
	- NEW Default User for connections based on device.
	- NEW Comments for Public Keys.

	- Security updates to libraries.

	- Fixed critical bugs affecting ssh connections.

# Version 1.031
	- NEW On-screen keyboard with more space for modifiers, FKeys and Cursor keys. Redesigned for more space on the modifiers, and with a central scrollable area that handles more keys. Activate the Alternate keys by taping on the Alt key. And now tap on a modifier to activate it as a normal button, or make a long press to chain different combinations.
	- NEW Add your own Fonts & Themes! More info on (https://github.com/blinksh/fonts) and (https://github.com/blinksh/themes)
	- NEW Multistep authentication. Servers with google authenticator or similar will now connect without problems :)
	- NEW Fira Code font with ligatures included.

	- Added -l parameter to ssh for Hosts.
	- Improved message on how use hosts after adding them.
	- Fixed on-screen arrows.
	- Fixed F0 as F10 on external keyboard.
	- Fixed keys selection problem on settings
	- Fixed ssh-copy-id issue when accessing the host for the first time

# Version 1.019
	- Simplified build process.
	- New README and BUILD instructions

# Version 0.1020
	- "Get In Touch" and "About" sections completed.
	- iOS 10 support.
	- Added exception for iOS distribution.

	- Improved gesture support, specially on iPhone and iOS10.
	- MBProgressHUD updated with patch.
	- Test Terminal fixes for fonts and updates.
	- Disable font ligatures.
	- SSH Session termination fixes.

# Version 0.1010
	- Terminal customization! Customize your terminal with different default themes and fonts. Preview your changes within the Settings Preview section. Save your changes and restore them on each execution. Enjoy!

	- Create an id_rsa on first boot as default key.

	- Swipe and pinch conflicts
	- Copy changes to settings

# Version 0.931
	- Map Caps or Shift taps to Esc key.
	- Cursor keys (Home, End, PgUp, PgDown) with default to Cmd+arrow.
	- Function keys (F1-F10) with default to Cmd+number.
	- Change mapping of any function key to a different combination.

	- Improved placeholders for Hosts configurations.

	- Removed Alt as Esc from initial defaults.

# Version 0.927
	This version should complete the experience in relation to ssh and terminal configurations, mimicking a big part of what you can do in a normal shell:
	- Hosts Configuration. Preconfigure a host parameters, like user, port, key and commands.
	- Connect to a Host by specifying its name. Do "mosh plankton"
	- Overwrite parameters from host configuration from the shell. So carlos@plankton will override the user field on plankton.
	- Default Modifier keys settings changed: Ctrl is Ctrl, and Alt sends Esc. Everything else is undefined.
	- Modifier keys configuration. CAPS as Ctrl or ESC, no problem! Cmd as Ctrl? You have it! Configure everything to you liking.
	- Added secure passwords stored on Keychain to Host Configuration.
	- Exiting the session within the MCP closes the Space.

	- Fixed hang after "exec request accepted". Establishing connections should be smooth now.
	- Fixed adjustments on viewport after rotating the display.
	- Fixed Ctrl + Space sequences.
	- Fixed wrong/unexisting Ctrl sequences.
	- Added Esc+any character support.

	- New settings added to project. Will start to fill up in raw branch.

# Version 0.916
	This version contains many important bug fixes that should improve the experience a lot and reduce crashes to, hopefully, close to zero:
	- Unicode support to Blink shell. Thanks Yury!
	- Support for function keys + modifiers. Now you can do Ctrl/Alt/Shift + Arrows.
	- Improved experience with drag down to close. Let me know what you think :)
	- SSH with improved function keys in ncurses apps. htop, less, etc... should now work perfectly.
	- New help message with version and gestures.
	- Added asterisk, underscore and escape to SmartKeys.
	- Added HockeySDKResources.bundle to project for improved updates.


AppleSource/shell_cmds-198/chroot/chroot.c
/*
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1988, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)chroot.c	8.1 (Berkeley) 6/9/93";
#else
__RCSID("$NetBSD: chroot.c,v 1.7 1998/10/06 03:47:51 mrg Exp $");
#endif
#endif /* not lint */

#include <sys/param.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <grp.h>
#include <paths.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int	main __P((int, char **));
void	usage __P((void)) __attribute__((__noreturn__));

char	*user;		/* user to switch to before running program */
char	*group;		/* group to switch to ... */
char	*grouplist;	/* group list to switch to ... */

int
main(argc, argv)
	int argc;
	char *argv[];
{
	struct group *gp;
	struct passwd *pw;
	char *shell, *endp, *comma;
	gid_t gid = 0, gidlist[NGROUPS_MAX];
	uid_t uid = 0;
	int ch, gids;

	while ((ch = getopt(argc, argv, "G:g:u:")) != -1)
		switch(ch) {
		case 'u':
			user = optarg;
			break;
		case 'g':
			group = optarg;
			break;
		case 'G':
			grouplist = optarg;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc < 1)
		usage();

	if (group) {
		if (isdigit(*group)) {
			gid = (gid_t)strtol(group, &endp, 0);
			if (endp == group)
				goto getgroup;
		} else {
getgroup:
			if ((gp = getgrnam(group)))
				gid = gp->gr_gid;
			else
				errx(1, "no such group %s", group);
		}
	}

	for (gids = 0; grouplist; ) {
		comma = strchr(grouplist, ',');

		if (comma)
			*comma++ = '\0';

		if (isdigit(*grouplist)) {
			gidlist[gids] = (gid_t)strtol(grouplist, &endp, 0);
			if (endp == grouplist)
				goto getglist;
		} else {
getglist:
			if ((gp = getgrnam(grouplist)))
				gidlist[gids] = gp->gr_gid;
			else
				errx(1, "no such group %s", group);
		}
		gids++;
		grouplist = comma;
	}

	if (user) {
		if (isdigit(*user)) {
			uid = (uid_t)strtol(user, &endp, 0);
			if (endp == user)
				goto getuser;
		} else {
getuser:
			if ((pw = getpwnam(user)))
				uid = pw->pw_uid;
			else
				errx(1, "no such user %s", user);
		}
	}

	if (chdir(argv[0]) || chroot("."))
		err(1, "%s", argv[0]);

	if (gids && setgroups(gids, gidlist) < 0)
		err(1, "setgroups");
	if (group && setgid(gid) < 0)
		err(1, "setgid");
	if (user && setuid(uid) < 0)
		err(1, "setuid");

	if (argv[1]) {
		execvp(argv[1], &argv[1]);
		err(1, "%s", argv[1]);
	}

	if (!(shell = getenv("SHELL")))
		shell = _PATH_BSHELL;
	execlp(shell, shell, "-i", NULL);
	err(1, "%s", shell);
	/* NOTREACHED */
}

void
usage()
{
	(void)fprintf(stderr, "usage: chroot [-g group] [-G group,group,...] "
	    "[-u user] newroot [command]\n");
	exit(1);
}
